pkg/cmd/pulumi/util_test.go

```diff
@@ -355,3 +356,122 @@ func TestStackLoadOption(t *testing.T) {
 		})
 	}
 }
+
+// TestGetUpdateMetadata tests that the update metadata is correctly populated
+// when running a Pulumi program.
+func TestPulumiCLIMetadata(t *testing.T) {
+	t.Parallel()
+
+	cmd := &cobra.Command{
+		Use: "my-cli-command",
+	}
+
+	cmd.PersistentFlags().String(
+		"name", "",
+		"This is a mock string flag that is set on the command.")
+
+	cmd.PersistentFlags().Int(
+		"age", 0,
+		"This is a mock int flag that is set on the command.")
+
+	cmd.PersistentFlags().Bool(
+		"human", false,
+		"This is a mock boolean flag that is set on the command.")
+
+	cmd.PersistentFlags().Bool(
+		"unused", false,
+		"This is an unused boolean flag that is NOT set on the command.")
+
+	actualEnv := map[string]string{}
+
+	getEnv := func() []string { // This is the function that returns the environment variables.
+		return []string{
+			// Checks that a normal flag is set in the environment.
+			"PULUMI_SKIP_UPDATE_CHECK=1",
+			// Checks that a flag is set in the environment to 'false'.
+			"PULUMI_EXPERIMENTAL=0",
+			// Checks that a non boolean flag set to true is 'set'.
+			"PULUMI_STRING_FLAG=true",
+			// Checks that an old and forgotten CLI flag is detected.
+			"PULUMI_DEPRECATED_FLAG=1.2.3",
+		}
+	}
+
+	subCmd := &cobra.Command{
+		Use: "subcommand",
+		Run: func(cmd *cobra.Command, args []string) {
+			addPulumiCLIMetadataToEnvironment(actualEnv, cmd.Flags(), getEnv)
+		},
+	}
+
+	cmd.AddCommand(subCmd)
+	cmd.SetArgs([]string{"subcommand", "--name", "pulumipus", "--age", "100", "--human"})
+
+	err := cmd.Execute()
+	assert.NoError(t, err, "Expected command to execute successfully")
+
+	// Check that normal flags are set in the environment.
+	for _, flagName := range []string{
+		"pulumi.version",
+		"pulumi.arch",
+		"pulumi.os",
+	} {
+		if flagName == "pulumi.version" {
+			// Skip the version check for now, since it's global and set by the compiler.
+			continue
+		}
+		value := actualEnv[flagName]
+		assert.NotEmpty(t, value, "Expected %s to be set in update environment map", flagName)
+	}
+
+	// Check that sensitive flags are not leaked in the environment.
+	for _, flag := range []struct {
+		key      string
+		expected string
+	}{
```

Comment:
Optional: This for second part could be a lot simpler if you delete the matched items in the loop above, and just do an assert.Equal:

```go
  assert.NotEmpty(t, value, "Expected %s to be set in update environment map", flagName)
  delete(actualEnv, flagName)
}

assert.Equal(t, map[string]string{
  // Command line flags:
  "pulumi.flag.name": "set",
  "pulumi.flag.age": "set",
  "pulumi.flag.human": "true",
  // Environment variables:
  "pulumi.env.PULUMI_SKIP_UPDATE_CHECK": "true",
  // ...
}, actualEnv)
```

it'll even correctly handle things that aren't expected in the map.